### PR TITLE
Tide: Set github status to success prior to merging

### PR DESCRIPTION
**Pull request by alvaroaleman**
This PR changes Tide to set its status context to success prior to merging. It is https://github.com/kubernetes/test-infra/pull/20892 plus a second commit to synchronize this with the status controller, ref https://github.com/kubernetes/test-infra/pull/20892#discussion_r578693503

Unblocks https://github.com/kubernetes/test-infra/issues/6289
Supersedes and thereby closes https://github.com/kubernetes/test-infra/pull/20892
Fixes https://github.com/kubernetes/test-infra/issues/16472
Fixes https://github.com/kubernetes/test-infra/issues/18782

/assign @cjwagner

### CODE REVIEW

File: prow/tide/status.go [hunk 1]
```diff
@@ -642,3 +647,36 @@ func contextCheckerGetterFactory(cfg *config.Config, gc git.ClientFactory, org,
 		return contextPolicy, nil
 	}
 }
+
+type minimalPR struct {
```
Comment (stevekuznetsov) [inline prow/tide/status.go:651]: nit: what is minimal about this? maybe `pullRequestIdentifier`?

File: prow/tide/tide.go [hunk 2]
```diff
@@ -408,6 +408,7 @@ func (c *Controller) Sync() error {
 	c.sc.requiredContexts = requiredContextsMap(filteredPools)
 	select {
 	case c.sc.newPoolPending <- true:
+		c.sc.dontUpdateStatus.reset()
```
Comment (stevekuznetsov) [inline prow/tide/tide.go:411]: What kind of guarantees do we have about the overall memory footprint of the set given how often we expect to call this?

Comment (alvaroaleman) [reply prow/tide/tide.go:411]: This set only contains PRs that we merged and is reset after status controller sync finished + tide sync finished it querying. The mem usage of this is absolutely negligible.

**Review by fejta — APPROVED** [review]
/hold

Worth putting this functionality behind a flag?

**Comment by alvaroaleman** [comment]
> Worth putting this functionality behind a flag?

I don't really think that is needed as the worst case scenario is that it doesn't work, which means we stay were we are today. And I would prefer to not increase the already big configuration surface further.

/hold cancel

**Pull request merged**